src/index/git_remote.rs

```diff
@@ -440,54 +441,71 @@ impl GitError {
+            }
+            _ => {}
         }
+
+        false
     }
 
     /// Returns true if a fetch could not be completed successfully due to the
     /// repo being locked, and could succeed if retried
     #[inline]
     pub fn is_locked(&self) -> bool {
-        match self {
-            Self::Fetch(gix::remote::fetch::Error::UpdateRefs(ure))
-            | Self::CloneFetch(gix::clone::fetch::Error::Fetch(
-                gix::remote::fetch::Error::UpdateRefs(ure),
-            )) => {
-                if let gix::remote::fetch::refs::update::Error::EditReferences(ere) = ure {
-                    match ere {
-                        gix::reference::edit::Error::FileTransactionPrepare(ftpe) => {
-                            use gix::refs::file::transaction::prepare::Error as PrepError;
-                            if let PrepError::LockAcquire { source, .. }
-                            | PrepError::PackedTransactionAcquire(source) = ftpe
-                            {
-                                // currently this is either io or permanentlylocked, but just in case
-                                // more variants are added, we just assume it's possible to retry
-                                // in anything but the permanentlylocked variant
-                                !matches!(
-                                    source,
-                                    gix::lock::acquire::Error::PermanentlyLocked { .. }
-                                )
-                            } else {
-                                false
-                            }
-                        }
-                        gix::reference::edit::Error::FileTransactionCommit(ftce) => {
-                            matches!(
-                                ftce,
-                                gix::refs::file::transaction::commit::Error::LockCommit { .. }
-                            )
-                        }
-                        _ => false,
-                    }
+        let ure = match self {
+            Self::Fetch(fe) => {
+                if let gix::remote::fetch::Error::UpdateRefs(ure) = &**fe {
+                    ure
+                } else {
+                    return false;
+                }
+            }
+            Self::CloneFetch(cf) => {
+                if let gix::clone::fetch::Error::Fetch(gix::remote::fetch::Error::UpdateRefs(ure)) =
+                    &**cf
+                {
+                    ure
                 } else {
-                    false
+                    return false;
                 }
             }
-            Self::Lock(le) => !matches!(le, gix::lock::acquire::Error::PermanentlyLocked { .. }),
-            _ => false,
+            Self::Lock(le) => {
+                return !matches!(le, gix::lock::acquire::Error::PermanentlyLocked { .. })
+            }
+            _ => return false,
+        };
+
+        if let gix::remote::fetch::refs::update::Error::EditReferences(ere) = ure {
```

Comment:
@NobodyXu I just wanted to point you at code that one would write currently in order to figure out specific error causes, and found it interesting to think about how it would be possible to streamline such *queries* with an error tree that is accessible more programmatically. The more I think about it, there more I believe we are onto something by trying to open another chapter in the "Rust Error Handling" book :).

Reply:
Thanks, if we are able to implement our proposal then it definitely would be a lot simpler.

P.S. I would have to do with binstall before implementing the proposal, I will start with cleaning up existing codebase by trying to remove the build.rs